javascript/fallback.js

```diff
@@ -0,0 +1,68 @@
+function fallbackWarning() {
+  console.warn(
+    "It seems like your environment does't support the native simsimd module, so we are providing a JS fallback."
+  );
+}
+
+function inner_distance(arr1, arr2) {
+  fallbackWarning();
+  if (arr1.length !== arr2.length) {
+    throw new Error("Vectors must have the same length");
+  }
+
+  return 1 - arr1.reduce((acc, val, i) => acc + val * arr2[i], 0);
+}
+
+function sqeuclidean(arr1, arr2) {
+  fallbackWarning();
+  if (arr1.length !== arr2.length) {
+    throw new Error("Vectors must have the same length");
+  }
+
+  return Math.hypot(...Object.keys(arr1).map((k) => arr2[k] - arr1[k])) ** 2;
+}
+
+function cosine(arr1, arr2) {
+  fallbackWarning();
+  if (arr1.length !== arr2.length) {
+    throw new Error("Vectors must have the same length");
+  }
+
+  const dotProduct = arr1.reduce((acc, val, i) => acc + val * arr2[i], 0);
+  const magnitudeA = Math.sqrt(arr1.reduce((acc, val) => acc + val ** 2, 0));
+  const magnitudeB = Math.sqrt(arr2.reduce((acc, val) => acc + val ** 2, 0));
+
```

Comment:
In your implementation, every array is passed twice. That should be more expensive than old-school `for`-loops over two arrays simultaneously. Also, unless the interpreter is very wise, it's always better to unpack low powers, like `x ** 2` into `x`. It will be much cheaper. I explain it in my [C++ benchmarking tutorial](https://github.com/ashvardanian/BenchmarkingTutorial/blob/fd548fae731e4273a723b4541c47417c78b689d1/tutorial.cxx#L58-L130).

Reply:
Yeah, originally I had it like that but then removed it for elegance sake.